### PR TITLE
New project show page :)

### PR DESCRIPTION
**Pull request by devton**
**Added components to mount project show:**
project_show

**project-highlight:**
project_highlight

**project-sidebar:**
captura de tela 2015-09-22 as 5 41 07 pm

**project-reminder:**
project_reminder

**project-share-box:**
project-share-box

**pop-notification:**
pop-notification

**project-nav:**
project-nav

**project-user-card:**
project-user-card

**project-tab:**
project-tab

**project-reward-list:**
project-reward-list

### CODE REVIEW

File: src/project/show.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+window.c.project.Show = (function(m, c, _, models, h){
+  return {
+    controller: function(args) {
+      var vm = m.postgrest.filtersVM({project_id: 'eq'}),
+          idVM = h.idVM,
+          projectDetails = m.prop([]),
+          userDetails = m.prop([]),
+          rewardDetails = m.prop([]);
+
+      vm.project_id(args.root.getAttribute('data-id'));
```
Comment (diogob) [inline src/project/show.js]: Let's add **id** to the project_details so we can always use ```h.idVM``` for the primary key

Comment (diogob) [inline src/project/show.js]: Let's use the permalink from the browser path to load the project

File: src/project/show.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
+window.c.project.Show = (function(m, c, _, models, h){
+  return {
+    controller: function(args) {
+      var vm = m.postgrest.filtersVM({project_id: 'eq'}),
```
Comment (diogob) [inline src/project/show.js]: let's use
```
var permalinkFilter = m.postgrest.filtersVM({permalink: 'eq'})
```

File: src/project/show.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        projectDetails(data);
+
+        idVM.id(project.user_id);
+        models.userDetail.getRowWithToken(idVM.parameters()).then(userDetails);
+      });
+
+      models.rewardDetail.getPage(vm.parameters()).then(rewardDetails);
+
+      return {
+        vm: vm,
```
Comment (diogob) [inline src/project/show.js]: we don't need to return this vm

File: src/project/show.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+window.c.project.Show = (function(m, c, _, models, h){
```
Comment (diogob) [inline src/project/show.js]: we need a spec for this

File: spec/components/project-highlight.spec.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+      $output = mq(view);
+    });
+
+    it('should render project video, headline, category and address info', () => {
+      expect($output.find('iframe.embedly-embed').length).toEqual(1);
+      expect($output.find('span.fa.fa-map-marker').length).toEqual(1);
+      expect($output.contains(projectDetail.address.city)).toEqual(true);
+    });
+
+    //FIXME: Pending spec
```
Comment (diogob) [inline spec/components/project-highlight.spec.js]: Let's fix this

File: src/c/project-tabs.js [hunk 6]
```diff
@@ -0,0 +1,31 @@
+window.c.ProjectTabs = (function(m, h){
```
Comment (diogob) [inline src/c/project-tabs.js]: we need a spec

File: src/c/project-main.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+              return routes['#about'];
+            }
+
+            return routes[hash];
+          };
+
+      window.addEventListener('hashchange', m.redraw, false);
+
+      return {
+        project: project,
```
Comment (diogob) [inline src/c/project-main.js]: we can remove this project

File: src/c/project-main.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
+window.c.ProjectMain = (function(m, c, _, h){
+  return {
+    controller: function(args) {
+      var project = args.project,
+          generateRoutes = function() {
```
Comment (diogob) [inline src/c/project-main.js]: let's use ```displayTabContent```

File: src/c/project-main.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+            var hash = window.location.hash,
+                c_opts = {project: project},
+                routes = {
```
Comment (diogob) [inline src/c/project-main.js]: let's use ```tabs```

File: src/c/project-contributions.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
+window.c.ProjectContributions = ((m, models, h, _) => {
+  return {
+    controller: (args) => {
+      let listVM = m.postgrest.paginationVM(models.projectContribution.getPageWithToken),
+          filterVM = m.postgrest.filtersVM({project_id: 'eq', waiting_payment: 'eq'}),
+          generateSort = (waiting = false) => {
+            return () => {
+              //FIXME: need to find a way to pass false filter
+              filterVM.waiting_payment(waiting);
+              listVM.firstPage(filterVM.parameters()).then(null);
```
Comment (diogob) [inline src/c/project-contributions.js]: ```then(null)``` ?

File: src/c/project-contributions.js [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+            };
+          };
+
+      filterVM.project_id(args.project.id);
+
+      if (!listVM.collection().length) {
+        listVM.firstPage(filterVM.parameters()).then(null);
```
Comment (diogob) [inline src/c/project-contributions.js]: ```then(null)``` ?

File: src/c/project-contributions.js [hunk 12]
```diff
@@ -0,0 +1,81 @@
+window.c.ProjectContributions = ((m, models, h, _) => {
+  return {
+    controller: (args) => {
+      let listVM = m.postgrest.paginationVM(models.projectContribution.getPageWithToken),
+          filterVM = m.postgrest.filtersVM({project_id: 'eq', waiting_payment: 'eq'}),
+          generateSort = (waiting = false) => {
```
Comment (diogob) [inline src/c/project-contributions.js]: why not ```toggleWaiting```

Comment (diogob) [inline src/c/project-contributions.js]: We should have a spec for this

File: src/c/project-about.js [hunk 13]
```diff
@@ -0,0 +1,29 @@
+window.c.ProjectAbout = (function(m, c, h){
```
Comment (diogob) [inline src/c/project-about.js]: we need a spec for this

File: src/c/project-posts.js [hunk 14]
```diff
@@ -0,0 +1,50 @@
+window.c.ProjectPosts = (function(m, models, h, _){
+  return {
+    controller: function(args) {
+      var listVM = m.postgrest.paginationVM(models.projectPostDetail.getPageWithToken),
+          filterVM = m.postgrest.filtersVM({project_id: 'eq'});
+
+      filterVM.project_id(args.project.id);
+
+      if (!listVM.collection().length) {
+        listVM.firstPage(filterVM.parameters()).then(null);
```
Comment (diogob) [inline src/c/project-posts.js]: let's remove the ```then(null)```

File: src/c/project-nav.js [hunk 15]
```diff
@@ -0,0 +1,88 @@
+window.c.ProjectNav = (function(m, _, h){
```
Comment (diogob) [inline src/c/project-nav.js]: need spec

Comment (diogob) [inline src/c/project-nav.js]: ```ProjectDashboardMenu``` ?

File: src/c/project-nav.js [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+              (!project.is_published ? '' : m('button#toggle-edit-menu.dashboard-nav-link-left', {onclick: ctrl.editLinksToggle.toggle}, [
+                m('span.fa.fa-pencil.fa-fw.fa-lg'), ' Editar projeto'
+              ])),
+              (ctrl.editLinksToggle() ? m('#edit-menu-items', [
+                m('#dashboard-links', [
+                  ((!project.is_published || project.is_admin_role) ? [
+                    m('a#basics_link[class="' + editLinkClass + '"][href="' + editRoute + '#basics' + '"]', 'Básico'),
+                    m('a#goal_link[class="' + editLinkClass + '"][href="' + editRoute + '#goal' + '"]', 'Meta e prazo'),
+                  ] : ''),
+                  m('a#description_link[class="' + editLinkClass + '"][href="' + editRoute + '#description' + '"]', 'Descrição'),
```
Comment (diogob) [inline src/c/project-nav.js]: maybe _.map([{
  video: "Vídeo",
  ...
}], (key, val) => {
m(`a#${key}_link[class="${editLinkClass}"][href="${editRoute}#${key}"]`, val)
})

File: src/c/project-share-box.js [hunk 17]
```diff
@@ -0,0 +1,44 @@
+window.c.ProjectShareBox = (function(m, h){
```
Comment (diogob) [inline src/c/project-share-box.js]: we need a spec

File: src/c/project-share-box.js [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+  return {
+    controller: function() {
+      var displayEmbed = h.toggleProp(false, true);
```
Comment (diogob) [inline src/c/project-share-box.js]: no need for var declaration, let's put directly in return:
```
return {
  displayEmbed: h.toggleProp(false, true)
};
```

File: src/c/project-share-box.js [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+            m('iframe[frameborder="0"][height="350px"][src="/projects/' + args.project.id + '/embed"][width="300px"][scrolling="no"]')
+          ])
+        ]) : ''),
+        m('a.w-hidden-main.w-hidden-medium.btn.btn-medium.btn-fb.u-marginbottom-20[href="http://www.facebook.com/sharer/sharer.php?u=https://www.catarse.me/' + args.project.permalink + '?ref=facebook&title=' + args.project.name + '"][target="_blank"]', [
+          m('span.fa.fa-facebook'),' Compartilhe'
+        ]),
+        m('a.w-hidden-main.w-hidden-medium.btn.btn-medium.btn-tweet.u-marginbottom-20[href="http://twitter.com/?status=Acabei de apoiar o projeto ' + args.project.name + ' htts://www.catarse.me/' + args.project.permalink + '?ref=twitterr"][target="_blank"]', [
+          m('span.fa.fa-twitter'),' Tweet'
+        ]),
+        //m('a.w-hidden-main.w-hidden-medium.btn.btn-medium[href=\'#\']', [
```
Comment (diogob) [inline src/c/project-share-box.js]: remove comments

File: src/c/project-share-box.js [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+  return {
+    controller: function() {
+      var displayEmbed = h.toggleProp(false, true);
+
+      return {
+        displayEmbed: displayEmbed
+      };
+    },
+    view: function(ctrl, args) {
+      return m('.pop-share', {style: 'display: block;'}, [
```
Comment (diogob) [inline src/c/project-share-box.js]: why display block ?

File: src/h.js [hunk 21]
```diff
@@ -53,15 +53,60 @@ window.c.h = (function(m, moment){
...
+    };
+
+    return window.setTimeout(tryParse, 500); //use timeout to wait async of facebook
+  },
+
+  pluralize = (count, s, p) => {
+    return (count > 1 ? count + p : count + s);
+  },
+
+  simpleFormat = (str) => {
```
Comment (diogob) [inline src/h.js]: we need a nice spec for this

File: src/h.js [hunk 22]
```diff
@@ -53,15 +53,60 @@ window.c.h = (function(m, moment){
...
+    str = str.replace(/\r\n?/, '\n');
+    if (str.length > 0) {
+      str = str.replace(/\n\n+/g, '</p><p>');
+      str = str.replace(/\n/g, '<br />');
+      str = '<p>' + str + '</p>';
+    }
+    return str;
+  },
+
+  rewardSouldOut = (reward) => {
```
Comment (diogob) [inline src/h.js]: spec

File: src/h.js [hunk 23]
```diff
@@ -53,15 +53,60 @@ window.c.h = (function(m, moment){
...
+    return (reward.maximum_contributions > 0 ?
+        (reward.paid_count + reward.waiting_payment_count >= reward.maximum_contributions) : false);
+  },
+
+  rewardRemaning = (reward) => {
```
Comment (diogob) [inline src/h.js]: spec

File: src/project/show.js (hunk 1 above)
Comment (devton) [inline src/project/show.js]: not sure, If we use permalink we need to create another idVM to handle with the project_id to request the reward list.

File: src/c/project-nav.js (hunk 16 above)
Comment (devton) [inline src/c/project-nav.js]: this does not work because im some cases we have badge

**Comment by devton** [comment]
@diogob has two specs failing because some weird stuff when we need to test .click interations. can you take a look :)

**Pull request merged**